docs/CONTRACTS_AND_TESTING.md

```diff
@@ -0,0 +1,899 @@
+
+### Basic Property Tests
+
+Use `@property` to define property-based tests:
+
+```cadenza
+@property
+fn prop_addition_commutative (a: Integer) (b: Integer) =
+  assert a + b == b + a
+
+@property
+fn prop_sort_idempotent (list: List Integer) =
+  let sorted_once = sort list
+  let sorted_twice = sort sorted_once
+  assert sorted_once == sorted_twice
+```
+
+### Property Test Configuration
+
+```cadenza
+@property { cases = 1000, max_size = 100 }
+fn prop_reverse_twice (list: List Integer) =
+  assert reverse (reverse list) == list
+
+@property { timeout = 5s }
+fn prop_no_infinite_loop (n: Integer) =
+  assert fibonacci n >= 0
+```
+
+### Shrinking
+
+When a property fails, automatically find the minimal failing case:
+
+```cadenza
+@property
+fn prop_all_positive (list: List Integer) =
+  assert all (map (x -> x > 0) list)
+
+# If this fails with list = [1, 2, -5, 10, 3], 
+# the shrinker will reduce it to list = [-5] or list = [0]
+# to show the minimal failing case
+```
+
+### Custom Generators
+
+Define how to generate arbitrary values for custom types:
+
+```cadenza
+# Automatic generator for simple types
+struct Point {
+  x = Float,
+  y = Float,
+}
+
+# The framework automatically generates arbitrary Points
+
+# Custom generator for constrained types
+type Email = String @ { is_valid_email x }
+
+@generator Email
+fn generate_email =
+  # Note: These are hypothetical helper functions for illustration
+  let username = generate_alphanumeric 10  # Generate 10 random alphanumeric chars
+  let domain = generate_domain             # Generate random domain like "example.com"
+  "${username}@${domain}"
+
+# Use in property tests
+@property
+fn prop_email_roundtrip (email: Email) =
+  assert parse_email (format_email email) == email
+```
+
+### Stateful Property Testing
+
+Test stateful systems with command sequences:
+
+```cadenza
+struct BankAccountModel {
+  balance = Float,
+}
+
+@stateful_property
+fn prop_bank_account_invariant =
+  let model = BankAccountModel { balance = 100.0 }
+  
+  @command "deposit"
+  fn deposit (amount: Float @ { x > 0.0 }) =
+    model.balance = model.balance + amount
+    assert model.balance >= 0.0
+  
+  @command "withdraw"
+  fn withdraw (amount: Float @ { x > 0.0 && x <= model.balance }) =
+    model.balance = model.balance - amount
+    assert model.balance >= 0.0
+
+# Framework generates random command sequences:
+# deposit 50.0, withdraw 30.0, deposit 20.0, withdraw 100.0, ...
+# Checks that invariant holds after each command
+```
+
+---
+
+## Static and Dynamic Analysis
+
+### Verification Levels
+
+Cadenza supports multiple levels of contract checking:
+
+1. **None**: Contracts are documentation only (fastest)
+2. **Dynamic**: Runtime checks (default)
+3. **Static**: Compile-time verification (slowest, most thorough)
+4. **Hybrid**: Static where possible, dynamic fallback
+
+### Configuration
+
+```cadenza
+# At module level
+@verification_level "static"
```

Comment:
I don't think we want this to be at the module level - it should be configured at the time of building.